pkg/controller/v1beta1/inferenceservice/reconcilers/knative/ksvc_reconciler.go

```diff
@@ -90,9 +90,12 @@ func createKnativeService(componentMeta metav1.ObjectMeta,
 	// so here we need to get the revision that is previously rolled out with 100%
 	if componentStatus.LatestRolledoutRevision == componentStatus.LatestReadyRevision &&
 		componentStatus.LatestReadyRevision == componentStatus.LatestCreatedRevision &&
-		componentExtension.CanaryTrafficPercent != nil && *componentExtension.CanaryTrafficPercent < 100 {
+		((componentExtension.CanaryTrafficPercent != nil && *componentExtension.CanaryTrafficPercent == 100) ||
+			(componentExtension.CanaryTrafficPercent != nil && *componentExtension.CanaryTrafficPercent == 0)) {
```

Comment:
Not sure if I understand the change, we want to only rollback to previous revision when canaryPercent is < 100% right?

Reply:
Exactly. But the process there is a bit chaotic. Let's take my situation above as an example. 

**When we roll out revision 03 with canary percent < 100%, it  will have 2 stages.**

- create revision 03 but revision 03 is not ready. 
- create revision 03 and revision 03 is ready.
In these two case, `lastestRolloutRevision` is 02. As in the second and third rows of the table above which below `trigger isvc canary 1% roll out revision03`
```
lastRolledoutRevision := componentStatus.LatestRolledoutRevision
// This is to handle case when the latest ready revision is rolled out with 100% and then rolled back
// so here we need to get the revision that is previously rolled out with 100%
if componentStatus.LatestRolledoutRevision == componentStatus.LatestReadyRevision &&
componentStatus.LatestReadyRevision == componentStatus.LatestCreatedRevision &&
((componentExtension.CanaryTrafficPercent != nil && *componentExtension.CanaryTrafficPercent == 100) ||
	(componentExtension.CanaryTrafficPercent != nil && *componentExtension.CanaryTrafficPercent == 0)) {
lastRolledoutRevision = componentStatus.PreviousRolledoutRevision
}
```
*latestRevision* is 03 and *preRevision* is 02 ,this if branch missed.

**When we roll out revision 03 with canary percent 100%**

As in the forth row of the table above. In this case we don't need rollback to previous, the if branch hit.

Then case when rolled out with 100% and then rolled back is similar to this.

It seems that the previous logic just happens to work, but it doesn't match the actual state, the BUG is caused by another suitation that when I roll out a new revision, a old componentStatus was wrongly triggered.

Reply:
@wenyangchou Thanks for the detailed explanation! I think you are right here, we can totally remove the condition. It incorrectly sets the traffic to `PreviousRolledoutRevision` before `LatestCreatedRevision` is updated in status when canarying a new revision. The correct state is set when [propagating the state](https://github.com/kserve/kserve/blob/master/pkg/apis/serving/v1beta1/inference_service_status.go), so removing this condition is fine and it just won't rollback the traffic immediately in the first reconciliation but once `latestRolledoutRevision` in status is rolled back the next reconciliation applies correctly.

Reply:
This can not totally remove the condition for the time being. Because in the scenario of roll out **the first revision** with traffic percent 1%, when `LatestRolledoutRevision` is not empty that means **revision 01** is ready. In this case the if condition should be hint, otherwise the `PreviousRolledoutRevision` would be the same as `LatestRolledoutRevision` instead of empty. As a result, the state of the isvc directly jumps repeatedly between true and false.

All in all, the if condition still needs to be retained.

Reply:
I am not sure if I follow here, the first revision should be always fully rolled out when there is only one revision. Also when `componentExtension.CanaryTrafficPercent == 100`, `lastRolledoutRevision` is not used anyway.

Reply:
Yeath, there are still some conditions missing. I'm adding.

Reply:
When you rollout a revision with `CanaryTrafficPercent` 0 you will hit the same problem as before where you can incorrectly bring up the previously rolled out revision.

Reply:
The if condition has been updated to 

```
if componentStatus.LatestRolledoutRevision == componentStatus.LatestReadyRevision &&
componentStatus.LatestReadyRevision == componentStatus.LatestCreatedRevision &&
((componentExtension.CanaryTrafficPercent != nil && *componentExtension.CanaryTrafficPercent == 100) ||
	(componentExtension.CanaryTrafficPercent != nil && *componentExtension.CanaryTrafficPercent == 0) ||
	componentStatus.PreviousRolledoutRevision == "") {
lastRolledoutRevision = componentStatus.PreviousRolledoutRevision
}
```
When I rollout a revision with CanaryTrafficPercent 0, it would not hit the if condition. For the `LastestRolloutRevision ` is not equals to the `LatestReadyRevision`. Only when the new revision has been rollout with traffic percent 100 can the `LastestRolloutRevision` be updated.

Reply:
```
componentStatus.PreviousRolledoutRevision == ""
```
another condition has been added.

Reply:
it is possible as the reconciliation can happen before `LatestCreatedRevision` and `LatestReadyRevision` in status are updated with the new revision, so you can still hit the if condition.

Reply:
The case I am talking about is when 
`LastestRolloutRevision`: 2, `LastestRolloutRevision`: 2, `LastestCreatedRevision`: 2
and then you update the isvc for some changes with 0 `CanaryTrafficPercent`, it creates a new revision but you can hit the if condition as they are still the same before status is updated with `LastestCreatedRevision`: 3.

Reply:
Emmm, I probably know what you said. After testing, it does.I will think about it.

Reply:
There seems to be no good way to solve this special case. Do you have any other good suggestions.

Reply:
These states are asynchronous and somewhat chaotic.

Reply:
I think removing the condition would solve it, for your special case I think we can probably add a validation that it does not allow setting canary traffic for the first revision(don't think there is a good reason to do so as well).

Reply:
see the events below for rolling out a revision with 0% then 100% and then rollback with 0%.
|action|LastestRollout|LatestReady|LatestCreated|PreviousRollout|Canary|Traffic split|
|-------------------------|--------------|--------------|--------------|--------------|--------------|-------|
|trigger isvc canary 0% roll out revision03|2|2|2|1|0|100(r2)|
|revision03 is created|2|2|3|1|0|100(r2)|
|revision03 is ready |2|3|3|1|0|100(r2)|
|update revision03 with 100% percent|2|3|3|1|100|100(r2)|
|revision03 traffic status updated to 100|3|3|3|2|100|100(r3)|
|revision03 roll back to 0|3|3|3|2|0|100(r3)|
|revision03 traffic status updated to 0|2|3|3|2|0|0(r3)/100(r2)|

Reply:
I think the if condition check is causing the chaos, we should be totally based on the state on the status.

Reply:
> I think removing the condition would solve it, for your special case I think we can probably add a validation that it does not allow setting canary traffic for the first revision(don't think there is a good reason to do so as well).

Well, there seems to be no better solution.